### PR TITLE
fix(ts): Stack overflow in Field metadata decoder for column-mapped tiles

### PR DESCRIPTION
**Pull request by Turtelll**
Bug
- The decoder crashed with a stack overflow error when decoding tiles with nesting/column mapping. 

Cause
- The decodeField() function in embeddedTilesetMetadataDecoder.ts was using an obsolete format that doesn't match how the encoding is done now. 
- When the Java encoder writes Field metadata for struct children, it follows this format: first a typeCode
  varint, then the field name string (if the typeCode indicates a name should be present), then the child count
   and children for complex types.
- TS still expected a old bitfield-based format

Fix
- alinges typescript decoder with java encoder format
- remove FieldOptions enum as it's not used anymore.

### CODE REVIEW

**Comment by HarelM** [comment]
Can you add a test to cover this change somehow? Either a tile with the relevant data or a unit test that simulates what happened.

**Comment by Turtelll** [comment]
I could hack together a test with a tile to test this but i would rather look to create proper integration tests that cover all encoding possibilities once fastPfor is added. The test fixtures we have aren't really good enough to do that. We should have a round trip decode-encode integration test that encoded a small tileset with all available options we have each time we run the full test suit. But I don't know if that is possible.

**Comment by HarelM** [comment]
For every bug we find it is best to write a test, this way the bug will not return as the test will fail if we change the code to reintroduce the bug...

**Comment by Turtelll** [comment]
@HarelM i added a tile with column mappings like name with multiple language variants (name:af, name:ar, name:az, name:be, etc.) this is exactly the nested struct/column mapping scenario that would trigger the stack overflow bug.

The test verifies:
  1. ✓ Tile decodes without throwing (no stack overflow)
  2. ✓ Metadata structure is valid (tables exist, have names, extent > 0)
  
I would still remove these kinds of tests once a sulution for general integration tests is implemented.

**Comment by HarelM** [comment]
Looks like the code you have changed is not fully covered.
Can you achieve the same coverage with a simple unit test instead of a full tile?
I would keep the tiles as a last resort, and as you said, a better tile covering strategy is underway, but I don't think it will solve all of our problems, I test to think that in a lot of cases unit tests are the key to really cover some edge cases...

**Comment by Turtelll** [comment]
@HarelM i removed the tile based test and added unit-tests

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+import { encodeVarintInt32Value } from "../../encoding/integerEncodingUtils";
+import { concatenateBuffers } from "../../decoding/decodingTestUtils";
+import { ComplexType, ScalarType } from "./tilesetMetadata";
+
+const STRUCT_TYPE_CODE = 30;
+
+/**
+ * Encodes a single typeCode as a varint.
+ */
+function encodeTypeCode(typeCode: number): Uint8Array {
```
Comment (HarelM) [inline ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts]: Can you move the encode functions to an encode file, much like the other encode methods?

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+function scalarTypeCode(scalarType: number, nullable: boolean): number {
+    return 10 + scalarType * 2 + (nullable ? 1 : 0);
+}
+
+describe("embeddedTilesetMetadataDecoder", () => {
+    describe("decodeField", () => {
+        describe("scalar fields", () => {
+            it.each([
+                {
```
Comment (HarelM) [inline ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts]: I think this is worth splitting to different tests, it doesn't save that much in terms of lines of code and it highlights the tests and improves readability.

**Comment by HarelM** [comment]
I've added a few minor comments.
Thanks for looking into this!

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts (hunk 1 above)
Comment (Turtelll) [reply ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts]: moved them to a new file in encoding folder

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts (hunk 2 above)
Comment (Turtelll) [reply ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts]: I split them up again

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+                const field = decodeField(buffer, new IntWrapper(0));
+
+                expect(field.name).toBe("address");
+                expect(field.nullable).toBe(false);
+                expect(field.type).toBe("complexField");
+                expect(field.complexField?.physicalType).toBe(ComplexType.STRUCT);
+                expect(field.complexField?.children).toHaveLength(children.length);
+
+                children.forEach((child, i) => {
```
Comment (HarelM) [inline ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts]: Sorry for the nit-picking, can you use regular "for" loop?

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+                const address = field.complexField?.children[0];
+                expect(address?.name).toBe("address");
+                expect(address?.type).toBe("complexField");
+
+                const coordinates = address?.complexField?.children[0];
+                expect(coordinates?.name).toBe("coordinates");
+                expect(coordinates?.complexField?.children).toHaveLength(leafChildren.length);
+
+                leafChildren.forEach((child, i) => {
```
Comment (HarelM) [inline ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts]: Same here - avoid using forEach.

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.ts [hunk 5]
```diff
@@ -36,54 +22,43 @@ function decodeString(src: Uint8Array, offset: IntWrapper): string {
...
+ */
+export function decodeField(src: Uint8Array, offset: IntWrapper): Field {
+    const typeCode = decodeVarintInt32(src, offset, 1)[0] >>> 0;
+    const column = decodeColumnType(typeCode);
 
-    const field = {} as Field;
-    if ((fieldOptions & FieldOptions.nullable) !== 0) {
-        field.nullable = true;
+    if (!column) {
+        throw new Error(`Unsupported field type code: ${typeCode}`);
```
Comment (HarelM) [inline ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.ts]: I think this error message should be elaborated.
For example, unsupported typecode x when trying to decode field 'something' , supported types are: a, b, c.
I'm not sure what part of this is possible, but the more descriptive the error is, the more chance a user will be able to really understand what went wrong.

These error types from MVT parsing are a good example of error code that don't inform the user what went wrong or how to fix the error they encountered.

**Comment by HarelM** [comment]
Almost there, added a few minor comments.

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts (hunk 3 above)
Comment (Turtelll) [reply ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.spec.ts]: sure

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.ts (hunk 5 above)
Comment (Turtelll) [reply ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.ts]: Changed the typeCode check to only scalar and struct, which decodable as fields. And tried to improve the error message.

File: ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.ts [hunk 6]
```diff
@@ -36,54 +25,43 @@ function decodeString(src: Uint8Array, offset: IntWrapper): string {
...
+ * Decodes a Field used as part of complex types (STRUCT children).
+ */
+export function decodeField(src: Uint8Array, offset: IntWrapper): Field {
+    const typeCode = decodeVarintInt32(src, offset, 1)[0] >>> 0;
+    const column = decodeColumnType(typeCode);
 
-    const field = {} as Field;
-    if ((fieldOptions & FieldOptions.nullable) !== 0) {
-        field.nullable = true;
+    if (typeCode <= 4) {
```
Comment (HarelM) [inline ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.ts]: Shouldn't this be 9 instead of 4? According to the message it only starts from 10...?

Comment (Turtelll) [reply ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.ts]: yes i was not finished. I need to adapt the test as well

**Comment by HarelM** [comment]
I've just noticed that the coverage on this file is not full, can you add coverage for the children code path?
https://app.codecov.io/gh/maplibre/maplibre-tile-spec/pull/790/blob/ts/src/metadata/tileset/embeddedTilesetMetadataDecoder.ts?dropdown=coverage

**Comment by HarelM** [comment]
image
